### PR TITLE
#162685668 Users can follow and unfollow

### PR DESCRIPTION
**Pull request by BarnaTB**
**What does this PR do?**
This PR enables users to follow each other's profiles and optionally unfollow each other.

**Description of Task to be completed?**
The application should enable a user to be able to follow another registered user's profile and unfollow them at will. The task should also enable a user to view their followers and other users that they're following.

**Link to pivotal tracker story**
[#162685668](https://www.pivotaltracker.com/story/show/162685668)

**How should this be manually tested?**

To follow a user, hit this endpoint, running a `PUT` request https://ah-backend-guardians-sta-pr-22.herokuapp.com/api/profile/follow/ and pass this model in the body.

Please note that you need to have registered this user first as it will give you an error message.

```
    {
        "username": "test"
    }
```

The output you should get back should be as in this **screenshot**.

image

To unfollow that user, just the request again and the output should appear as in the screenshot below.

image

To view all your followers, hit the endpoint https://ah-backend-guardians-sta-pr-22.herokuapp.com/api/profile/followers/ , running a `GET`request and you should get output that looks as in the **screenshot**.

image

To view all the people you're following hit the endpoint https://ah-backend-guardians-sta-pr-22.herokuapp.com/api/profile/following/ , running a `GET`request and you should get back output as in the **screenshot**.

image

### CODE REVIEW

**Review by fahadmak — CHANGES_REQUESTED** [review]
Please use a descriptive error message when trying to follow a user who does not exist instead of

`IndexError at /api/profile/follow/`
`list index out of range`

**Comment by BarnaTB** [comment]
> Please use a descriptive error message when trying to follow a user who does not exist instead of
> 
> `IndexError at /api/profile/follow/`
> `list index out of range`

@fahadmak I've fixed the issue.

**Review by fahadmak — COMMENTED** [review]
please use pe8 to standardize your code

**Review by fahadmak — APPROVED** [review]

File: authors/apps/profiles/tests/test_views.py [hunk 1]
```diff
@@ -57,7 +55,57 @@ def test_unauthorized_to_edit_profile(self):
...
+    def test_view_following(self):
+        """Method to test that a user can view all the users they're following
+        """
+        self.client.force_authenticate(user=self.user)
+        response = self.client.put(self.url, data=follow_user, format='json')
+        self.assertEqual(response.data['message'],
+                         'You have followed bgpeter!')
+        self.client.force_authenticate(user=self.user)
+        response = self.client.get(self.url3)
+        print(response.data)
```
Comment (VivianDoreen) [inline authors/apps/profiles/tests/test_views.py]: kindly remove the print statement from the codebase

File: authors/apps/profiles/views.py [hunk 2]
```diff
@@ -3,58 +3,131 @@
...
+class FollowProfileView(generics.UpdateAPIView):
+    """View class to handle user's follower operations
+    """
+    permission_classes = (IsAuthenticated,)
+
+    def put(self, request):
+        """Method to follow a user's profile
+        """
+        username = request.data.get('username', {})
+        print(username)
```
Comment (VivianDoreen) [inline authors/apps/profiles/views.py]: kindly remove the print statement

**Review by VivianDoreen — CHANGES_REQUESTED** [review]
Kindly use descriptive error messages for followers who don't exist and if the username filled is empty


image

image

image

image

**Review by fahadmak — CHANGES_REQUESTED** [review]
please check for these handle errors descriptively
screen shot 2019-01-22 at 8 28 49 am
screen shot 2019-01-22 at 9 04 18 am
screen shot 2019-01-22 at 9 04 40 am
screen shot 2019-01-22 at 9 04 53 am

**Comment by BarnaTB** [comment]
@fahadmak and @VivianDoreen please use the review app as explained in the PR. This is happening because you're testing using users from the old application which did not have profiles. However, with current developments, all users automatically get profiles when they register so you might want to register new users to test the feature. This is also explained in the PR.

**Review by fahadmak — APPROVED** [review]

**Review by VivianDoreen — APPROVED** [review]

File: authors/apps/profiles/views.py [hunk 3]
```diff
@@ -68,3 +52,92 @@ class ListProfileView(generics.ListAPIView):
...
+            0]['id']
+        queryset = Profile.objects.get(user__pk=user_id).followers.all()
+
+        followers = []
+        for follower in queryset:
+            follower_name = User.objects.get(email=follower.email)
+            follower_username = follower_name.username
+            followers.append(follower_username)
+        if followers is None or followers == []:
+            msg = {'error': 'No followers were found!'}
```
Comment (archbishop12) [inline authors/apps/profiles/views.py]: @BarnaTB Change this to message because it is not an error when one has no followers.
```suggestion
            msg = {'message': 'No followers were found!'}
```

File: authors/apps/profiles/views.py [hunk 4]
```diff
@@ -68,3 +52,92 @@ class ListProfileView(generics.ListAPIView):
...
+            0]['id']
+        queryset = User.objects.filter(pk=user_id).first().is_following.all()
+
+        following = []
+        for follower in queryset:
+            following_name = User.objects.get(username=follower)
+            following_username = following_name.username
+            following.append(following_username)
+        if following is None or following == []:
+            msg = {'error': 'No one is following you yet!'}
```
Comment (archbishop12) [inline authors/apps/profiles/views.py]: @BarnaTB Change the key here to message. It is not an error when you are not following anyone.
```suggestion
            msg = {'message': 'You are not following anyone!'}
```

File: .env.example [hunk 5]
```diff
@@ -18,7 +18,7 @@ EMAIL_HOST_PASSWORD = email_host_password
 
 EMAIL_PORT = EMAIL_PORT
 
-cloud_name = <your-cloud-name>
+CLOUD_NAME = <your-cloud-name>
```
Comment (archbishop12) [inline .env.example]: @BarnaTB This is an unnecessary change , you should leave it like before.

File: requirements.txt [hunk 6]
```diff
@@ -4,8 +4,6 @@ autopep8==1.4.3
 certifi==2018.11.29
 cffi==1.11.5
 chardet==3.0.4
-coreapi==2.3.3
```
Comment (archbishop12) [inline requirements.txt]: @BarnaTB Please add the packages you removed here , some are required for the api documentation.

**Review by archbishop12 — APPROVED** [review]
LGTM @BarnaTB I am merging this.
